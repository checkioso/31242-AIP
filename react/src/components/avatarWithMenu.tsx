import React, { useState } from "react";
import { Link as RouterLink } from "react-router-dom";
import {
  Avatar,
  Link,
  ListItemIcon,
  Menu,
  MenuItem,
  MenuProps,
  withStyles,
  ListItemText,
} from "@material-ui/core";
import {
  MeetingRoom,
  PeopleOutline,
  ThumbsUpDown,
  ExitToApp,
} from "@material-ui/icons";

// kudos to https://material-ui.com/components/menus/
const StyledMenu = withStyles({
  paper: {
    border: "1px solid #d3d4d5",
  },
})((props: MenuProps) => (
  <Menu
    elevation={0}
    getContentAnchorEl={null}
    anchorOrigin={{
      vertical: "bottom",
      horizontal: "center",
    }}
    transformOrigin={{
      vertical: "top",
      horizontal: "center",
    }}
    {...props}
  />
));

// kudos to https://material-ui.com/components/menus/
const StyledMenuItem = withStyles((theme) => ({
  root: {
    "&:focus": {
      backgroundColor: theme.palette.primary.main,
      "& .MuiListItemIcon-root, & .MuiListItemText-primary": {
        color: theme.palette.common.white,
      },
    },
  },
}))(MenuItem);

type AvatarWithMenuProps = {
  fullName: string;
};

// most readable solution from https://stackoverflow.com/questions/33076177/getting-name-initials-using-js
function nameToUpperInitials(fullName: string) {
  const namesArray = fullName.split(" ");
  if (namesArray.length === 1)
    return `${namesArray[0].charAt(0).toUpperCase()}`;
  else
    return `${namesArray[0].charAt(0).toUpperCase()}${namesArray[
      namesArray.length - 1
    ]
      .charAt(0)
      .toUpperCase()}`;
}

function AvatarWithMenu(props: AvatarWithMenuProps) {
  const [userMenu, setUserMenu] = useState<boolean>(false);
  const [anchorEl, setAnchorEl] = useState<HTMLElement | null>(null);

  return (
    <div>
      <Avatar
        onClick={(event: React.MouseEvent<HTMLElement>) => {
          setUserMenu(true);
          setAnchorEl(event.currentTarget);
        }}
        id="avatar"
      >
        {nameToUpperInitials(props.fullName)}
      </Avatar>
      <StyledMenu
        id="customized-menu"
        anchorEl={anchorEl}
        keepMounted
        open={userMenu}
        onClose={() => {
          setUserMenu(false);
          setAnchorEl(null);
        }}
      >
        <Link
          color="inherit"
          component={RouterLink}
          to="/"
          style={{ textDecoration: "none" }}
        >
          <StyledMenuItem>
            <ListItemIcon>
              <MeetingRoom fontSize="small" />
            </ListItemIcon>
            <ListItemText primary="Click here to go back!" />
          </StyledMenuItem>
        </Link>
        <Link
          id="favoursLink"
          color="inherit"
          component={RouterLink}
          to={{ pathname: "/user", state: { tabIndex: 1 } }}
          style={{ textDecoration: "none" }}
        >
          <StyledMenuItem>
            <ListItemIcon>
              <ThumbsUpDown fontSize="small" />
            </ListItemIcon>
            <ListItemText primary="My Favours" />
          </StyledMenuItem>
        </Link>
        <Link
          id="requestLink"
          color="inherit"
          component={RouterLink}
          to={{ pathname: "/user", state: { tabIndex: 2 } }}
          style={{ textDecoration: "none" }}
        >
          <StyledMenuItem>
            <ListItemIcon>
              <PeopleOutline fontSize="small" />
            </ListItemIcon>
            <ListItemText primary="My Requests" />
          </StyledMenuItem>
        </Link>
        <MenuItem />
        <Link
          id="logOutLink"
          color="inherit"
          component={RouterLink}
          to="/login"
          style={{ textDecoration: "none" }}
        >
          <StyledMenuItem>
            <ListItemIcon>
              <ExitToApp fontSize="small" />
            </ListItemIcon>
            <ListItemText primary="Log Out" />
          </StyledMenuItem>
        </Link>
      </StyledMenu>
    </div>
  );
}

export { AvatarWithMenu, nameToUpperInitials };
