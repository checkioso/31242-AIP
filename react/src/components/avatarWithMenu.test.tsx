import React from "react";
import { shallow } from "enzyme";
import { AvatarWithMenu } from "./avatarWithMenu";

// Integration tests are on home.test.tsx
describe("<AvatarWithMenu />", () => {
  it("should render correctly", () => {
    const wrapper = shallow(
      <AvatarWithMenu loggedIn={true} fullName={"Ben Johnston"} />
    );

    expect(wrapper).toMatchSnapshot();
  });
});
