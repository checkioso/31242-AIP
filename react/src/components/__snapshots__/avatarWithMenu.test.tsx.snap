// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AvatarWithMenu /> should render correctly 1`] = `
<div>
  <WithStyles(ForwardRef(Avatar))
    id="avatar"
    onClick={[Function]}
  >
    BJ
  </WithStyles(ForwardRef(Avatar))>
  <WithStyles(Component)
    anchorEl={null}
    id="customized-menu"
    keepMounted={true}
    onClose={[Function]}
    open={false}
  >
    <WithStyles(ForwardRef(Link))
      color="inherit"
      component={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "displayName": "Link",
          "propTypes": Object {
            "innerRef": [Function],
            "onClick": [Function],
            "replace": [Function],
            "target": [Function],
            "to": [Function],
          },
          "render": [Function],
        }
      }
      style={
        Object {
          "textDecoration": "none",
        }
      }
      to="/"
    >
      <WithStyles(WithStyles(ForwardRef(MenuItem)))>
        <WithStyles(ForwardRef(ListItemIcon))>
          <MeetingRoomIcon
            fontSize="small"
          />
        </WithStyles(ForwardRef(ListItemIcon))>
        <WithStyles(ForwardRef(ListItemText))
          primary="Click here to go back!"
        />
      </WithStyles(WithStyles(ForwardRef(MenuItem)))>
    </WithStyles(ForwardRef(Link))>
    <WithStyles(ForwardRef(Link))
      color="inherit"
      component={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "displayName": "Link",
          "propTypes": Object {
            "innerRef": [Function],
            "onClick": [Function],
            "replace": [Function],
            "target": [Function],
            "to": [Function],
          },
          "render": [Function],
        }
      }
      id="favoursLink"
      style={
        Object {
          "textDecoration": "none",
        }
      }
      to={
        Object {
          "pathname": "/user",
          "state": Object {
            "tabIndex": 1,
          },
        }
      }
    >
      <WithStyles(WithStyles(ForwardRef(MenuItem)))>
        <WithStyles(ForwardRef(ListItemIcon))>
          <ThumbsUpDownIcon
            fontSize="small"
          />
        </WithStyles(ForwardRef(ListItemIcon))>
        <WithStyles(ForwardRef(ListItemText))
          primary="My Favours"
        />
      </WithStyles(WithStyles(ForwardRef(MenuItem)))>
    </WithStyles(ForwardRef(Link))>
    <WithStyles(ForwardRef(Link))
      color="inherit"
      component={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "displayName": "Link",
          "propTypes": Object {
            "innerRef": [Function],
            "onClick": [Function],
            "replace": [Function],
            "target": [Function],
            "to": [Function],
          },
          "render": [Function],
        }
      }
      id="requestLink"
      style={
        Object {
          "textDecoration": "none",
        }
      }
      to={
        Object {
          "pathname": "/user",
          "state": Object {
            "tabIndex": 2,
          },
        }
      }
    >
      <WithStyles(WithStyles(ForwardRef(MenuItem)))>
        <WithStyles(ForwardRef(ListItemIcon))>
          <PeopleOutlineIcon
            fontSize="small"
          />
        </WithStyles(ForwardRef(ListItemIcon))>
        <WithStyles(ForwardRef(ListItemText))
          primary="My Requests"
        />
      </WithStyles(WithStyles(ForwardRef(MenuItem)))>
    </WithStyles(ForwardRef(Link))>
    <WithStyles(ForwardRef(MenuItem)) />
    <WithStyles(ForwardRef(Link))
      color="inherit"
      component={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "displayName": "Link",
          "propTypes": Object {
            "innerRef": [Function],
            "onClick": [Function],
            "replace": [Function],
            "target": [Function],
            "to": [Function],
          },
          "render": [Function],
        }
      }
      id="logOutLink"
      style={
        Object {
          "textDecoration": "none",
        }
      }
      to="/login"
    >
      <WithStyles(WithStyles(ForwardRef(MenuItem)))>
        <WithStyles(ForwardRef(ListItemIcon))>
          <ExitToAppIcon
            fontSize="small"
          />
        </WithStyles(ForwardRef(ListItemIcon))>
        <WithStyles(ForwardRef(ListItemText))
          primary="Log Out"
        />
      </WithStyles(WithStyles(ForwardRef(MenuItem)))>
    </WithStyles(ForwardRef(Link))>
  </WithStyles(Component)>
</div>
`;
